cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailureDetectorTest.java

```diff
@@ -86,62 +95,81 @@ public void testFailureDetection() throws Exception {
+    BrokerFailureDetector detector = createBrokerFailureDetector(anomalies, mockTime, failedBrokersFile.getPath());
     try {
+      String failedBrokerListString = detector.loadPersistedFailedBrokerList();
+      assertTrue(failedBrokerListString.isEmpty());
       int brokerId = 0;
+      long anomalyTime = mockTime.milliseconds();
       killBroker(brokerId);
       detector.startDetection();
       assertEquals(Collections.singletonMap(brokerId, 100L), detector.failedBrokers());
+      failedBrokerListString = detector.loadPersistedFailedBrokerList();
+      assertEquals(String.format("%d=%d", brokerId, anomalyTime), failedBrokerListString);
     } finally {
       detector.shutdown();
+      Files.delete(failedBrokersFile.toPath());
     }
   }
 
   @Test
-  public void testLoadFailedBrokersFromZK() throws Exception {
+  public void testLoadFailedBrokersFromFile() throws Exception {
     Time mockTime = getMockTime();
     Queue<Anomaly> anomalies = new PriorityBlockingQueue<>(ANOMALY_DETECTOR_INITIAL_QUEUE_SIZE, anomalyComparator());
-    BrokerFailureDetector detector = createBrokerFailureDetector(anomalies, mockTime);
-
+    File failedBrokersFile = File.createTempFile("testLoadFailedBrokersFromFile", ".txt");
+    BrokerFailureDetector detector = createBrokerFailureDetector(anomalies, mockTime, failedBrokersFile.getPath());
     try {
+      String failedBrokerListString = detector.loadPersistedFailedBrokerList();
+      assertTrue(failedBrokerListString.isEmpty());
       detector.startDetection();
       int brokerId = 0;
+      long anomalyTime = mockTime.milliseconds();
       killBroker(brokerId);
       long start = System.currentTimeMillis();
       while (anomalies.isEmpty() && System.currentTimeMillis() < start + 30000) {
         // Wait for the anomalies to be drained.
       }
-      assertEquals(Collections.singletonMap(brokerId, 100L), detector.failedBrokers());
+      assertEquals(Collections.singletonMap(brokerId, anomalyTime), detector.failedBrokers());
       // shutdown, advance the clock and create a new detector.
       detector.shutdown();
+
       mockTime.sleep(100L);
-      detector = createBrokerFailureDetector(anomalies, mockTime);
+      detector = createBrokerFailureDetector(anomalies, mockTime, failedBrokersFile.getPath());
+      failedBrokerListString = detector.loadPersistedFailedBrokerList();
+      assertEquals(String.format("%d=%d", brokerId, anomalyTime), failedBrokerListString);
       // start the newly created detector and the broker down time should remain previous time.
       detector.startDetection();
-      assertEquals(Collections.singletonMap(brokerId, 100L), detector.failedBrokers());
+      assertEquals(Collections.singletonMap(brokerId, anomalyTime), detector.failedBrokers());
+      failedBrokerListString = detector.loadPersistedFailedBrokerList();
+      assertEquals(String.format("%d=%d", brokerId, anomalyTime), failedBrokerListString);
```

Comment:
Can we add another test case for migrating CC to a different host, i.e., failedBrokersFile is not present in the new host?

Reply:
Migrating CC to a different host without carrying `_failedBrokersFile` would not do any fatal harm to internal CC cluster model. It would cause CC to not detect the initial failure time of broker(s), which failed in the middle of the grace period that CC waits before taking action on the failed brokers. As a result, during such a migration, CC might wait longer than intended as a side effect -- which should be a rare event.